Refactor up/down moves

heathhenley opened the pull request:
Attempt to close #17

heathhenley committed:
Refactory up/down moves

heathhenley committed:
Add guard for negatives

chshersh commented:
I propose to focus on the other two PRs first and get them merged:

- #19 
- #20 

Once they're done, I'll review this one 🙂

chshersh commented:
Since the other PRs are merged, I propose to update this one, and I'll review it

heathhenley commented:
Just rebased on the latest `main`

chshersh commented on lib/fs/fs.ml:
```diff
@@ -70,29 +65,35 @@ let zipper_parents zipper =
+  let new_cursor = move_cursor move_dir move_file old in
+  { zipper with current = new_cursor }
+
+let go_down zipper = go_move 1 zipper
+let go_up zipper = go_move (-1) zipper
```
Beautiful ✨ 

We can even do eta-reduction and make the code slightly more concise
```suggestion
let go_down = go_move 1
let go_up = go_move (-1)
```

chshersh commented on lib/fs/fs.ml:
```diff
@@ -70,29 +65,35 @@ let zipper_parents zipper =
-  | File_cursor cursor ->
-      let new_offset = max 0 (Filec.offset cursor - 1) in
-      { zipper with current = update_cursor cursor new_offset }
+let move_cursor move_dir_cursor move_file_cursor cursor =
+  match cursor with
```
_Nit_: Since we're working on the cursor as the last argument, we can use the `function` syntax to avoid redundant names and get to the point faster.
```suggestion
let move_cursor move_dir_cursor move_file_cursor = function
```

chshersh commented on lib/fs/fs.ml:
```diff
@@ -70,29 +65,35 @@ let zipper_parents zipper =
+  let new_pos = (cursor.pos + move + len) mod len in
+  { cursor with pos = new_pos }
+
+let move_file_cursor move cursor =
+  let len = Filec.length cursor in
+  let new_offset = Filec.offset cursor + move in
+  let new_offset = max 0 new_offset in
+  if new_offset + span > len then cursor
+  else
+    match cursor with
+    | Text cur -> Text { cur with offset = new_offset }
+    | Binary -> cursor
```
Let's inverse the order of matching a little bit here.

Let's match on `cursor` first and if it's `Binary`, return immediately. Otherwise, proceed with the rest.

OCaml is eager and there's no point in do all these calculations for the `Binary` case.

heathhenley replied:
That make sense!

One issue I ran into that I don't really fully understand - when we match first and pull out the `Filec.Text` variant, we can't use our module functions on it because they expect a `Filec.t` - I really expected them to work because it's one of the options for `Filec.t` - but it gives this error if you try the use the value unpacked from the match (eg `txt_cur` in the latest version) vs using a `File.t` in for example:  `let cur = Filec.Text txt_cur`:
```
Error: This form is not allowed as the type of the inlined record could escape.
``` 

Is the approach in (5fe3a10)  a reasonable way to  handle that? Or should the `Filec` module functions be able to be called on the "unpacked" versions? Eg they are currently for example:

``` ocaml
(** Returns the len of file contents **)
val length : t -> int
```
but should there also be public functions like:
``` ocaml
(** Returns the len of file contents **)
val length : Text -> int
```
etc for the others?

Maybe if so, the ` t -> int`  (using `length` as an example) would call the appropriate `Text -> int` or `Binary -> int` ones in the match?

chshersh replied:
Ah, it's a bit annoying issue 😞 

Probably a sign that the current design is not perfect and we face such awkward moments. I don't see an elegant fix immediately, so let's keep as is and think about better ways in background since the affected code is not huge.

chshersh reviewed (commented):
Great stuff!

I left a bunch of nitpicks, mostly some stylistic improvements.

After looking at the diff, it may seem like this refactoring doesn't do a lot. But in the future, there'll be lots of zipper-like and scrollable elements, so it makes sense to prepare the ground for this work and start abstracting some of these movements.

heathhenley committed:
Apply suggestions from code review

Readability / style changes

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

heathhenley committed:
Swap order to save work

chshersh commented on lib/fs/fs.ml:
```diff
@@ -70,29 +65,35 @@ let zipper_parents zipper =
+let move_file_cursor move cursor =
+  match cursor with
+  | Filec.Binary -> cursor
+  | Filec.Text txt_cur ->
+      let cur = Filec.Text txt_cur in
+      let len = Filec.length cur in
+      let new_offset = Filec.offset cur + move in
+      let new_offset = max 0 new_offset in
+      if new_offset + span > len then cur
+      else Filec.Text { txt_cur with offset = new_offset }
```
This is really just 7 lines of code but they can be quite difficult to understand when forgetting the context 😅 

Let's do a few minor refactorings:

1. Remove the local variable `let cur = ...`. We can use `cursor` instead since it has the same type. One variable less!
2. The following code basically just keeps the current cursor if we go out of bounds but it's written in a slightly convoluted way (my bad, I probably was tired to write it properly).

```ocaml
let new_offset = max 0 new_offset in
if new_offset + span > len then cur
```

Let's write it more explicitly like this:

```ocaml
if new_offset < 0 || new_offset + span > len then cur
else Filec.Text { txt_cur with offset = new_offset }
```

heathhenley committed:
Switch to pipe and make logic more succinct

heathhenley commented:
I just rebased on the latest `main` and added the suggestion you had above to clean up the offset check logic. I also rewrote to use pipes but not sure if it's actually better here, lmk what you think!

chshersh commented on lib/fs/fs.ml:
```diff
@@ -70,29 +65,33 @@ let zipper_parents zipper =
+  | Filec.Binary -> cursor
+  | Filec.Text txt_cur ->
+      let len = Filec.length cursor in
+      let new_offset = move |> ( + ) (Filec.offset cursor) |> max 0 in
```
_Nit:_ All good, except we can simplify this part because in the next line we check `if new_offset < 0`, so there's no need to do `max 0` anymore
```suggestion
      let new_offset = Filec.offset cursor + move in
```

heathhenley replied:
🤦‍♂️

heathhenley committed:
Update lib/fs/fs.ml

Co-authored-by: Dmitrii Kovanikov <kovanikov@gmail.com>

chshersh reviewed (approved):
Thanks for the refactoring!
